source/simulator/nullspace.cc

```diff
@@ -165,6 +165,88 @@ namespace aspect
 
   }
 
+
+
+  template <int dim>
+  void Simulator<dim>::setup_nullspace_constraints(ConstraintMatrix &constraints)
+  {
+    if (!(parameters.nullspace_removal & (NullspaceRemoval::linear_momentum
+                                          | NullspaceRemoval::net_translation)))
+      return;
+
+    const unsigned int flags[] = {(NullspaceRemoval::linear_momentum_x
+                                   |NullspaceRemoval::net_translation_x),
+                                  (NullspaceRemoval::linear_momentum_y
+                                   |NullspaceRemoval::net_translation_y),
+                                  (NullspaceRemoval::linear_momentum_z
+                                   |NullspaceRemoval::net_translation_z)
+                                 };
+
+    // First find candidates for DoF indices to constrain for each velocity component.
+    types::global_dof_index vel_idx[dim];
+    {
+      for (unsigned int d=0; d<dim; ++d)
+        vel_idx[d] = numbers::invalid_dof_index;
+
+      unsigned int n_left_to_find = dim;
+
+      std::vector<types::global_dof_index> local_dof_indices (finite_element.dofs_per_cell);
+      typename DoFHandler<dim>::active_cell_iterator cell;
+      for (cell = dof_handler.begin_active(); n_left_to_find>0 && cell != dof_handler.end(); ++cell)
+        if (cell->is_locally_owned())
+          {
+            cell->get_dof_indices (local_dof_indices);
+
+            for (unsigned int i=0; i<finite_element.dofs_per_cell; ++i)
+              {
+                const unsigned int component = finite_element.system_to_component_index(i).first;
```

Comment:
Ah, excellent point. That's why you need the loop over all dofs (and, in fact, over more than possibly just the first locally owned cell -- maybe a comment to this effect would be a good idea).

The thing with system_to_component_index vs component_to_system index was that you try to identify only x-velocity dofs. You can do this with the first of these functions by looking at all dofs and finding out which of those are x-velocities. Or you can simply loop over all x-velocity dofs and then using the second of the two functions to find out the how-manyth dof they correspond to on the local cell.

As far as performance is concerned, it won't matter -- you do this at most once per time step, on at most a small number of cells. It was just a matter of making the code easy to understand, but I'm not feeling particular about it.